### PR TITLE
Backport 271078 to staging 23.11

### PR DESCRIPTION
**Pull request by SomeoneSerge**
## Description of changes

Supersedes https://github.com/NixOS/nixpkgs/pull/272336. Incorporates #272063, #273048, https://github.com/NixOS/nixpkgs/pull/272964 and https://github.com/NixOS/nixpkgs/pull/272812

This probably goes into staging before https://github.com/NixOS/nixpkgs/pull/272784

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/compilers/cudatoolkit/hooks/mark-for-cudatoolkit-root-hook.sh [hunk 1]
```diff
@@ -1,8 +1,14 @@
 # shellcheck shell=bash
 
+# Should we mimick cc-wrapper's "hygiene"?
+[[ -z ${strictDeps-} ]] || (( "$hostOffset" < 0 )) || return 0
```
Comment (SomeoneSerge) [inline pkgs/development/compilers/cudatoolkit/hooks/mark-for-cudatoolkit-root-hook.sh:4]: I shouldn't backport this bit though

Comment (ConnorBaker) [reply pkgs/development/compilers/cudatoolkit/hooks/mark-for-cudatoolkit-root-hook.sh:4]: @SomeoneSerge why not?

File: pkgs/development/libraries/nvidia-optical-flow-sdk/default.nix [hunk 2]
```diff
@@ -18,6 +18,11 @@ stdenv.mkDerivation {
     cp -R * $out/include
   '';
 
+  postFixup = ''
+    mkdir -p $out/nix-support
+    echo $pname >> "$out/nix-support/include-in-cudatoolkit-root"
+  '';
+
```
Comment (ConnorBaker) [inline pkgs/development/libraries/nvidia-optical-flow-sdk/default.nix:25]: This feels a tiny bit like magic -- is it possible to get the same effect by including the CUDA setup hook?

Comment (SomeoneSerge) [reply pkgs/development/libraries/nvidia-optical-flow-sdk/default.nix:25]: Damn, this is exactly what `cudaPackages.markForCudatoolkitRootHook` does 🤦🏻 this already went to master though

Comment (ConnorBaker) [reply pkgs/development/libraries/nvidia-optical-flow-sdk/default.nix:25]: Would you mind submitting a fix for master and cherry-picking it here?

File: pkgs/development/compilers/cudatoolkit/hooks/mark-for-cudatoolkit-root-hook.sh (hunk 1 above)
Comment (jonringer) [reply pkgs/development/compilers/cudatoolkit/hooks/mark-for-cudatoolkit-root-hook.sh:4]: it's potentially breaking behavior change  where this logic ran regardless before.

Comment (SomeoneSerge) [reply pkgs/development/compilers/cudatoolkit/hooks/mark-for-cudatoolkit-root-hook.sh:4]: I attached the comment at the wrong file btw, This one only affects `cudaPackages.markForCudatoolkitRootHook` which is used internally in `cudaPackages`, so it's OK to keep it. I had commented out the guard in the `setupCudaHook` already

File: pkgs/development/libraries/nvidia-optical-flow-sdk/default.nix (hunk 2 above)
Comment (SomeoneSerge) [reply pkgs/development/libraries/nvidia-optical-flow-sdk/default.nix:25]: Yes, I should

**Review by ConnorBaker — APPROVED** [review]

**Comment by ConnorBaker** [comment]
Just waiting for `ofborg-eval` to finish before merging...

**Comment by vcunat** [comment]
At a glance, I'd say that such large changes aren't even suitable for backporting to stable.

**Comment by SomeoneSerge** [comment]
Cctag definitely was affected by the opencvConfig issue on master at the time, as were two dozen other cudaSupport=true packages in the release branch (link tbd, writing from a mobile). The backport is huge partly because I tried to merge and reorder fixes from several correlated PRs. I'll look into the hydra failure, thx

**Comment by vcunat** [comment]
For reference, target PR #279139

**Pull request merged**